distributed/shuffle/_disk.py

```diff
@@ -65,38 +76,62 @@ async def _process(self, id: str, shards: list[bytes]) -> None:
         dropping the write into communicate above.
         """
 
+        # Normalisation for safety
+        id = str(id)
         with log_errors():
             # Consider boosting total_size a bit here to account for duplication
             with self.time("write"):
-                with open(
-                    self.directory / str(id), mode="ab", buffering=100_000_000
-                ) as f:
-                    for shard in shards:
-                        f.write(shard)
+                if not self.max_in_memory_buffer_size:
+                    # Fast path if we're always hitting the disk
+                    self._write(id, shards)
+                else:
+                    while shards:
+                        if self._in_memory < self.max_in_memory_buffer_size:
+                            self._memory_buf[id].append(newdata := shards.pop())
+                            self._in_memory += len(newdata)
+                        else:
+                            # Flush old data
+                            # This could be offloaded to a background
+                            # task at the cost of going further over
+                            # the soft memory limit.
+                            for k, v in self._memory_buf.items():
+                                self._write(k, v)
+                            self._memory_buf.clear()
+                            self._in_memory = 0
+
+    def _write(self, id: str, shards: list[bytes]) -> None:
+        with open(self.directory / str(id), mode="ab", buffering=100_000_000) as f:
+            for s in shards:
+                f.write(s)
+
+    async def close(self) -> None:
+        await super().close()
+        with contextlib.suppress(FileNotFoundError):
+            shutil.rmtree(self.directory)
 
     def read(self, id: int | str) -> bytes:
-        """Read a complete file back into memory"""
+        """Read a complete file back into memory, concatting with any
+        in memory parts"""
         self.raise_on_exception()
         if not self._inputs_done:
             raise RuntimeError("Tried to read from file before done.")
 
+        id = str(id)
```

Comment:
Should probably promote stringification to the caller.

Reply:
yes, had the same thought a couple of times already but was never worth the effort

Reply:
I think the callers all make strings, so I just removed this and tightened the type.